### ISSUE TITLE
Error BASE/1302 Object destructor failure: Reference to freed block

### ISSUE THREAD

**Issue by JoseQuintas** [harbour/core#107]
Receiving the error after this commit
https://github.com/vszakats/harbour-core/commit/46fe44162be5f9b119cc82ed0d368ac181c99074

Using Harbour 3.4dev
Application is 32 bits
S.O. is 64 bits
compiler msvc 2010
GTWVG
Error occurs on Windows Server 2008, but not on Windows 7 Ultimate

I do not know if is code dependent:

```
Error BASE/1302 Object destructor failure: Reference to freed block
Called from HB_IDLESLEEP(0)
Called from MAIN(76) 
```

```harbour
   RunThread( { || Sistema( xParam ) } )
   DO WHILE RunThread()
      HB_IdleSleep(1)
   ENDDO
. . . 
FUNCTION RunThread( bCode )

   STATIC aThreadList := {}, s_Mutex := hb_MutexCreate(), nTries := 0
   LOCAL  nCont, lIsRunning := .F.

   hb_MutexLock( s_Mutex )

   IF bCode == NIL
      IF Len( aThreadList ) != 0
         lIsRunning := .F.
         FOR nCont = Len( aThreadList ) TO 1 STEP -1
            IF hb_ThreadWait( aThreadList[ nCont ], 0.1, .T. ) != 1
               lIsRunning := .T.
            ELSE
               aDel( aThreadList, nCont )
               aSize( aThreadList, Len( aThreadList ) - 1 )
            ENDIF
         NEXT
      ENDIF
   ELSE
      Aadd( aThreadList, hb_ThreadStart( bCode ) )
      lIsRunning := .T.
      Inkey(0.5) // to wait thread load before thread kill routine
   ENDIF
   IF lIsRunning
      nTries := 0
   ELSE
      nTries += 1
      IF nTries < 3
         lIsRunning := .T.
      ENDIF
   ENDIF
   hb_MutexUnlock( s_Mutex )
   RETURN lIsRunning
```

**Comment by vszakats** [comment]
Original report here: https://github.com/vszakats/harbour-core/issues/172
(It doesn't seem related to any fork difference - where core C codebase is practically identical.)
Commit: 46fe44162be5f9b119cc82ed0d368ac181c99074

**Comment by JoseQuintas** [comment]
Test with change https://github.com/harbour/core/commit/5eca8eab9227100a9aadb42bdd4c556038a3a620
Ok. Thanks.

**Comment by JoseQuintas** [comment]
I made a new test, using Windows 2016 Server Preview, and get a GPF

```
Unrecoverable error 6005: Exception error:

Exception Code:C0000005 ACCESS_VIOLATION
Exception Address:00936E68
EAX:0000000A EBX:00000000 ECX:00008000 EDX:00C9C760
ESI:00000034 EDI:0106B874 EBP:00F0FEB8
CS:EIP:0023:00936E68 SS:ESP:002B:00F0FD88
DS:002B ES:002B FS:0053 GS:002B
Flags:00010206
Exception Parameters: 00000000 00000008
CS:EIP: 0F B7 48 FE 8B D1 83 E2 FD 66 3B 15 E8 82 D5 00
SS:ESP: 009376E8 0106C5AC 5CED6A0F 0106B574 0093790D 0106B874 0090AB00 0106B854 0093B97E 5CED6A0F 0106B854 00F0FEB8 00C4CE18 00940718 00000000 0000C118

C stack:
EIP: EBP: Frame: OldEBP, RetAddr, Params...
00936E68 00F0FEB8 00F0FF80 008D6630 00C4CD30 00D49490 009228C3 00000000 00000000 00000000 00000000 0000004C
008D6630 00F0FF80 00F0FF94 746C3484 7FA43000 746C3460 FA113CB4
746C3484 00F0FF94 00F0FFDC 775EF165 7FA43000 F90EE357 00000000 00000000 7FA43000 C0000005 7697AB80 00000000
775EF165 00F0FFDC 00F0FFEC 775EF130 FFFFFFFF 7760A914


Modules:
008C0000 0054B000 C:\JPA\HAROLDO\jpa\JPA.exe
77590000 00179000 C:\Windows\SYSTEM32\ntdll.dll
746B0000 000F0000 C:\Windows\SYSTEM32\KERNEL32.DLL
76840000 00178000 C:\Windows\SYSTEM32\KERNELBASE.dll
744A0000 00091000 C:\Windows\system32\apphelp.dll
76C60000 0007B000 C:\Windows\SYSTEM32\ADVAPI32.dll
76B40000 000BE000 C:\Windows\SYSTEM32\msvcrt.dll
76140000 00042000 C:\Windows\SYSTEM32\sechost.dll
76340000 000AC000 C:\Windows\SYSTEM32\RPCRT4.dll
745B0000 0001E000 C:\Windows\SYSTEM32\SspiCli.dll
745A0000 0000A000 C:\Windows\SYSTEM32\CRYPTBASE.dll
74540000 0005A000 C:\Windows\SYSTEM32\bcryptPrimitives.dll
763F0000 000BE000 C:\Windows\SYSTEM32\COMDLG32.dll
77370000 001BA000 C:\Windows\SYSTEM32\combase.dll
76DE0000 0008C000 C:\Windows\SYSTEM32\shcore.dll
76580000 00140000 C:\Windows\SYSTEM32\USER32.dll
74810000 0014D000 C:\Windows\SYSTEM32\GDI32.dll
76C00000 00044000 C:\Windows\SYSTEM32\SHLWAPI.dll
74960000 013BD000 C:\Windows\SYSTEM32\SHELL32.dll
76E80000 004E1000 C:\Windows\SYSTEM32\windows.storage.dll
76570000 0000C000 C:\Windows\SYSTEM32\kernel.appcore.dll
74660000 00044000 C:\Windows\SYSTEM32\powrprof.dll
747A0000 0000F000 C:\Windows\SYSTEM32\profapi.dll
75FB0000 000EA000 C:\Windows\SYSTEM32\ole32.dll
760A0000 00092000 C:\Windows\SYSTEM32\OLEAUT32.dll
747B0000 0005C000 C:\Windows\SYSTEM32\WS2_32.dll
75D50000 00007000 C:\Windows\SYSTEM32\NSI.dll
74290000 00209000 C:\Windows\WinSxS\x86_microsoft.windowscommon-controls_6595b64144ccf1df_6.0.10514.0_none_d3c0e55165dc1043\COMCTL32.dll
74220000 00067000 C:\Windows\SYSTEM32\WINSPOOL.DRV
74200000 0001B000 C:\Windows\SYSTEM32\bcrypt.dll
75D20000 0002B000 C:\Windows\SYSTEM32\IMM32.DLL
769C0000 00120000 C:\Windows\SYSTEM32\MSCTF.dll
741F0000 00006000 C:\Windows\system32\msimg32.dll
74170000 00074000 C:\Windows\system32\uxtheme.dll
74150000 0001D000 C:\Windows\system32\dwmapi.dll
76CF0000 00082000 C:\Windows\SYSTEM32\clbcatq.dll
74040000 0010F000 C:\Program Files (x86)\Common Files\System\ado\msado15.dll
74020000 00020000 C:\Windows\SYSTEM32\MSDART.DLL
73F50000 000CE000 C:\Program Files (x86)\Common Files\System\Ole DB\oledb32.dll
73F40000 00008000 C:\Windows\SYSTEM32\DPAPI.dll
73DF0000 0014E000 C:\Windows\SYSTEM32\comsvcs.dll
73D50000 0009C000 C:\Program Files (x86)\Common Files\System\Ole DB\msdasql.dll
73C90000 0001D000 C:\Program Files (x86)\Common Files\System\Ole DB\MSDATL3.dll
73CB0000 00099000 C:\Windows\SYSTEM32\ODBC32.dll
04610000 00010000 C:\Program Files (x86)\Common Files\System\Ole DB\MSDASQLR.DLL
73C70000 00013000 C:\Windows\SYSTEM32\cryptsp.dll
73C40000 0002F000 C:\Windows\system32\rsaenh.dll
737D0000 0046A000 C:\Program Files (x86)\MySQL\Connector ODBC 3.51\myodbc3.dll
737C0000 0000A000 C:\Windows\SYSTEM32\Secur32.dll
737B0000 00008000 C:\Windows\SYSTEM32\WSOCK32.dll
73790000 00012000 C:\Windows\system32\napinsp.dll
73740000 0004E000 C:\Windows\System32\mswsockdll
736B0000 00084000 C:\Windows\SYSTEM32\DNSAPI.dll
736A0000 0000B000 C:\Windows\System32\winrnr.dll
73680000 00014000 C:\Windows\system32\NLAapi.dll
73670000 00008000 C:\Windows\System32\rasadhlp.dll
73640000 00030000 C:\Windows\SYSTEM32\IPHLPAPI.DLL
73630000 00008000 C:\Windows\SYSTEM32\WINNSI.DLL
735E0000 00047000 C:\Windows\System32\fwpuclnt.dll
73540000 00099000 C:\Program Files (x86)\Common Files\System\msadc\msadce.dll
04C80000 00003000 C:\Program Files (x86)\Common Files\System\msadc\msadcer.dll

Called from HB_IDLESLEEP(0)
Called from MAIN(76)
```
reverting the 2 commits ok.

**Comment by vszakats** [comment]
Please reopen the Issue in mainline and post the new GPF there. Thanks! Closing this to track this under a single Issue.

**Comment by druzus** [comment]
In above I do not see any relation to main code.
If you can find it then please send self contain example with exact main line changelog entry which introduced the problem in new new issue.

**Comment by vszakats** [comment]
(Deleted my prev comment - I thought the follow-up was posted in the fork Issue, but we're here in mainline already. Sorry)

**Comment by JoseQuintas** [comment]
You are right.
When I get error with last version and back to same version ok in W2008 it runs ok.
I consider same commit, but can be any of about 50 commits.
I will do more tests.

**Comment by JoseQuintas** [comment]
Solved. Change hb_IdleState(.5) to Inkey(.5) in main thread.
May be W2016 consider main thread sleeping as a freeze application.

**Comment by vszakats** [comment]
Avoiding a GPF by replacing `hb_idleState()` with `Inkey()` doesn't seem to be resolving the root cause.

**Comment by JoseQuintas** [comment]
Can be a stupid question, but is possible that Windows or ADO or thread need any information of main thread, and the Idle State cause a different answer?

**Comment by vszakats** [comment]
APIs may have certain thread requirements, so it's not a stupid question. But, here we're talking about an unreleased Windows version under development, so anything can happen in it, it may even be an OS problem. It'd nevertheless be interesting to inspect _where_ exactly (which component's which function/line) the GPF happens.

**Comment by JoseQuintas** [comment]
Today when turn on W2016, it installs a update.
Then, start tests again, installing a new W2016 machine.

A strange behaviour:

from my machine, I copy a folder to my server \\serverjpa\d\newfolder
Acessing from my machine, I can see the new folder
from W2016, the folder becomes visible only after several seconds.
From prompt same thing.

Now, I am not sure about any test on W2016, could be any EXE.

I think this is part of the off-line work.
